jax/lax/lax_control_flow.py

```diff
@@ -1532,23 +1563,37 @@ def masked(*args):
   const_avals, carry_avals, x_avals = split_list(jaxpr.in_avals, [num_consts, num_carry])
   return _make_typed_jaxpr(masked, [aval] + const_avals + [aval] + carry_avals + x_avals)
 
-def scan_bind(*args, reverse, length, num_consts, num_carry, jaxpr, linear):
-  if not core.skip_checks:
-    assert len(linear) == len(args)
-    consts, init, xs = split_list(args, [num_consts, num_carry])
-    consts_avals, init_avals, x_avals = split_list(jaxpr.in_avals, [num_consts, num_carry])
-    assert all(_map(typecheck, consts_avals, consts)), (consts, consts_avals)
-    assert all(_map(typecheck, init_avals, init))
-    carry_avals, _ = split_list(jaxpr.out_avals, [num_carry])
-    assert all(_map(typematch, init_avals, carry_avals))
-    core.check_jaxpr(jaxpr.jaxpr)
-  return core.Primitive.bind(scan_p, *args, reverse=reverse, length=length,
-                             jaxpr=jaxpr, num_consts=num_consts,
-                             num_carry=num_carry, linear=linear)
+def _scan_typecheck(*avals, reverse, length, num_consts, num_carry, jaxpr,
```

Comment:
Can you add calls to this function in the transformation rules, guarded by `core.skip_checks`? The reason to have the checks in `scan_bind` was to catch bugs in transform rules as early as possible. Unless we add calls to `_scan_typecheck`, we'll be removing a lot of checks.

Similar for `_cond_typecheck`.

Reply:
Good call!